From 9b50dfb2dabd7465824f5cf229017d328f05f52e Mon Sep 17 00:00:00 2001
From: Jean-Philippe Brucker <jean-philippe@linaro.org>
Date: Wed, 9 Feb 2022 18:22:26 +0000
Subject: [PATCH 60/97] KVM: arm64: pkvm: Add a SMMUv3 driver

Add the skeleton for an Arm SMMUv3 driver at EL2.

Signed-off-by: Jean-Philippe Brucker <jean-philippe@linaro.org>
---
 arch/arm64/kvm/hyp/nvhe/Makefile            |  2 ++
 arch/arm64/kvm/hyp/nvhe/iommu/arm-smmu-v3.c | 26 +++++++++++++++
 arch/arm64/kvm/hyp/nvhe/setup.c             |  3 +-
 include/kvm/arm_smmu_v3.h                   | 36 ++++++++++++++++++++-
 4 files changed, 65 insertions(+), 2 deletions(-)
 create mode 100644 arch/arm64/kvm/hyp/nvhe/iommu/arm-smmu-v3.c

diff --git a/arch/arm64/kvm/hyp/nvhe/Makefile b/arch/arm64/kvm/hyp/nvhe/Makefile
index ad262c45d8fd..6f8b69e39d96 100644
--- a/arch/arm64/kvm/hyp/nvhe/Makefile
+++ b/arch/arm64/kvm/hyp/nvhe/Makefile
@@ -26,6 +26,8 @@ obj-y += $(lib-objs)
 
 obj-$(CONFIG_KVM_S2MPU) += iommu/s2mpu.o
 
+obj-$(CONFIG_ARM_SMMU_V3_PKVM) += iommu/arm-smmu-v3.o
+
 ##
 ## Build rules for compiling nVHE hyp code
 ## Output of this folder is `kvm_nvhe.o`, a partially linked object
diff --git a/arch/arm64/kvm/hyp/nvhe/iommu/arm-smmu-v3.c b/arch/arm64/kvm/hyp/nvhe/iommu/arm-smmu-v3.c
new file mode 100644
index 000000000000..7fce2dbcc7db
--- /dev/null
+++ b/arch/arm64/kvm/hyp/nvhe/iommu/arm-smmu-v3.c
@@ -0,0 +1,26 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * pKVM hyp driver for the Arm SMMUv3
+ *
+ * Copyright (C) 2022 Linaro Ltd.
+ */
+#include <asm/kvm_hyp.h>
+#include <kvm/arm_smmu_v3.h>
+
+size_t __ro_after_init kvm_hyp_arm_smmu_v3_nr_devices;
+struct hyp_arm_smmu_v3_device __ro_after_init *kvm_hyp_arm_smmu_v3_devices;
+
+static int smmu_init(void)
+{
+	return -ENOSYS;
+}
+
+static struct kvm_iommu_ops smmu_ops = {
+	.init				= smmu_init,
+};
+
+int kvm_arm_smmu_v3_register(void)
+{
+	kvm_iommu_ops = smmu_ops;
+	return 0;
+}
diff --git a/arch/arm64/kvm/hyp/nvhe/setup.c b/arch/arm64/kvm/hyp/nvhe/setup.c
index 955eb1b37097..7faaed59d80d 100644
--- a/arch/arm64/kvm/hyp/nvhe/setup.c
+++ b/arch/arm64/kvm/hyp/nvhe/setup.c
@@ -19,6 +19,7 @@
 #include <nvhe/mm.h>
 #include <nvhe/pkvm.h>
 #include <nvhe/trap_handler.h>
+#include <kvm/arm_smmu_v3.h>
 
 unsigned long hyp_nr_cpus;
 
@@ -346,7 +347,7 @@ static int select_iommu_ops(enum kvm_iommu_driver driver)
 	case KVM_IOMMU_DRIVER_NONE:
 		return 0;
 	case KVM_IOMMU_DRIVER_SMMUV3:
-		return -ENOSYS;
+		return kvm_arm_smmu_v3_register();
 	}
 
 	return -EINVAL;
diff --git a/include/kvm/arm_smmu_v3.h b/include/kvm/arm_smmu_v3.h
index 23132a6d13b3..4c27b5b51445 100644
--- a/include/kvm/arm_smmu_v3.h
+++ b/include/kvm/arm_smmu_v3.h
@@ -2,10 +2,44 @@
 #ifndef __KVM_ARM_SMMU_V3_H
 #define __KVM_ARM_SMMU_V3_H
 
+#include <asm/kvm_asm.h>
+
+#if IS_ENABLED(CONFIG_ARM_SMMU_V3_PKVM)
+
+struct hyp_arm_smmu_v3_device {
+	/* Parameters */
+	phys_addr_t		mmio_addr;
+	size_t			mmio_size;
+};
+
+extern size_t kvm_nvhe_sym(kvm_hyp_arm_smmu_v3_nr_devices);
+#define kvm_hyp_arm_smmu_v3_nr_devices \
+	kvm_nvhe_sym(kvm_hyp_arm_smmu_v3_nr_devices)
+
+extern struct hyp_arm_smmu_v3_device *kvm_nvhe_sym(kvm_hyp_arm_smmu_v3_devices);
+#define kvm_hyp_arm_smmu_v3_devices \
+	kvm_nvhe_sym(kvm_hyp_arm_smmu_v3_devices)
+
+#endif /* CONFIG_ARM_SMMU_V3_PKVM */
+
+/*
+ * Hypervisor interface
+ */
+#ifdef __KVM_NVHE_HYPERVISOR__
+# if IS_ENABLED(CONFIG_ARM_SMMU_V3_PKVM)
+int kvm_arm_smmu_v3_register(void);
+
+# else /* CONFIG_ARM_SMMU_V3_PKVM */
+static inline int kvm_arm_smmu_v3_register(void)
+{
+	return -EINVAL;
+}
+# endif /* CONFIG_ARM_SMMU_V3_PKVM */
+
 /*
  * Host interface
  */
-#ifndef __KVM_NVHE_HYPERVISOR__
+#else /* !__KVM_NVHE_HYPERVISOR__ */
 # if IS_ENABLED(CONFIG_ARM_SMMU_V3_PKVM)
 int kvm_arm_smmu_v3_init(void);
 void kvm_arm_smmu_v3_remove(void);
-- 
2.34.1

